Migrate the Test Explorer to use FSAutoComplete / VSTest

farlee2121 opened the pull request:
# Test Explorer VSTest Migration

Depends on [FSAutoComplete PR 1383](https://github.com/ionide/FsAutoComplete/pull/1383)

## Motivation

- [Microsoft is releasing a new test platform, and it doesn't support the `dotnet test` parameter we use](https://github.com/ionide/ionide-vscode-fsharp/issues/2069)
- [Access to standardized test code locations](https://github.com/ionide/ionide-vscode-fsharp/pull/2000) / The same code location support as VS
- Ability to write automated tests against test explorer code
- Ability to share test explorer features across FSAC clients
- [Improve test discovery speed](https://github.com/ionide/ionide-vscode-fsharp/issues/2038)

## Before this PR

The Ionide test explorer run on dotnet test and TRX files. 
Test discovery and test execution are implemented directly in Ionide.

## After this PR

Test discovery and execution are moved to the language server, allowing them to be used across clients.
Users can revert to the dotnet test-based approach with the setting `FSharp.TestExplorer.UseLegacyDotnetCliIntegration`

When using the LSP discovery
- we now support Microsoft.Testing.Platform VSTestBridge projects
- We have access to test code locations provided by the test libraries themselves
- We don't have to run tests to discover them, which makes test discovery much faster
- We don't need to create temporary artifacts (i.e. TRX files)
- VSTest manages test execution across assemblies, reducing our responsibility for managing resource utilization

We will not yet support Microsoft Testing Platform projects that don't use VSTestBridge.

## Outstanding issues

- I tried to support .runsettings ([ionide 2093](https://github.com/ionide/ionide-vscode-fsharp/discussions/2093) and [2040](https://github.com/ionide/ionide-vscode-fsharp/discussions/2040)), but there are [outstanding questions](https://github.com/ionide/FsAutoComplete/pull/1383#issuecomment-3258623066) on how to support it via VSTestConsoleWrapper.

- Microsoft Testing Platform VSTestBridge projects report their test locations as the end of a test's code block.
There's not really a way for us to fix this.

- Additional work is needed to support pure Microsoft Testing Platform projects.

- We'll probably want to remove the dotnet test-based code once we're confident the LSP-based holds up to user expectations

- There is some duplication of logic between Ionide and FSAutoComplete, specifically around deciding what projects are test projects


I also considered moving more logic to the language server for better testing, but I decided to hold off.
The current progress is useful, and supporting Microsoft.Testing.Platform is likely to add new constraints to the design.

farlee2121 commented on src/Components/TestExplorer.fs:
```diff
@@ -1525,19 +1988,108 @@ module Interactions =
+                else
+                    do! runTests_WithLanguageServer mergeTestResultsToExplorer testController.items req testRun
+                    testRun.``end`` ()
+                    do! discoverTests_WithLangaugeServer testItemFactory testController.items tryGetLocation
```
Test runs can add test to the explorer. but the test run will not provide a code location.
This makes for a confusing and bad experience.

I tried a number of alternatives, and this one felt the best. 
Running after the test run avoids the test run feeling sluggish, and the discovery still runs fast enough (even with thousands of tests) that it's generally done by the time a dev is ready for a new action.

[More information](https://github.com/ionide/ionide-vscode-fsharp/pull/2106/commits/9c97dae6ab0e91edef4d2d247b5e991e3a943296)

A key alternative here would be running test discovery whenever a the test project build artifact is updated.
I think that's more in line with what Visual Studio does, but was a bigger deviation from our existing workflow

farlee2121 commented on src/Components/TestExplorer.fs:
```diff
@@ -1469,10 +1735,208 @@ module Interactions =
+                    showStarted progress.ActiveTests
+                    mergeResults TrimMissing.NoTrim progress.TestResults
+
+                    let appendTestResultToOutput (testResult: TestResultDTO) =
```
I found printing all the test results to the test output was helpful feedback. I felt more confident that the test run went as expected or could see any error messages.

I could see an argument that printing all the test results feels noisy for big test runs.

farlee2121 commented on src/Components/TestExplorer.fs:
```diff
@@ -1433,8 +1705,7 @@ module Interactions =
                 |> Option.defaultWith (fun () ->
-                    let message =
-                        $"Could not run tests: project does not found in workspace. {projectPath}"
+                    let message = $"Could not run tests: project not found in workspace. {projectPath}"
```
The grammar error was already fixed. Copilot seems to be confused by the before half of the diff

TheAngryByrd commented:
i'm going to test over the next few days. I'll leave feedback as I go.


When I ran a test a second time against IcedTasks I started getting an error:

```
[20:21:44 DEBUG] [TestExplorer] TestRunRequest hp {
  include: [
    MP {
      id: 'c:\\Users\\jimmy\\Repositories\\public\\TheAngryByrd\\IcedTasks2\\tests\\IcedTasks.Tests.NS21\\IcedTasks.Tests.NS21.fsproj -- IcedTasks.AsyncEx.AsyncExBuilder.Return',
      uri: [Ia],
      children: [Object],
      range: [Getter/Setter],
      label: [Getter/Setter],
      description: [Getter/Setter],
      sortText: [Getter/Setter],
      canResolveChildren: [Getter/Setter],
      busy: [Getter/Setter],
      error: [Getter/Setter],
      tags: [Getter/Setter]
    },
    MP {
      id: 'c:\\Users\\jimmy\\Repositories\\public\\TheAngryByrd\\IcedTasks2\\tests\\IcedTasks.Tests.NS20\\IcedTasks.Tests.NS20.fsproj -- IcedTasks.AsyncEx.AsyncExBuilder.Return',
      uri: [Ia],
      children: [Object],
      range: [Getter/Setter],
      label: [Getter/Setter],
      description: [Getter/Setter],
      sortText: [Getter/Setter],
      canResolveChildren: [Getter/Setter],
      busy: [Getter/Setter],
      error: [Getter/Setter],
      tags: [Getter/Setter]
    },
    MP {
      id: 'c:\\Users\\jimmy\\Repositories\\public\\TheAngryByrd\\IcedTasks2\\tests\\IcedTasks.Tests\\IcedTasks.Tests.fsproj -- IcedTasks.AsyncEx.AsyncExBuilder.Return',
      uri: [Ia],
      children: [Object],
      range: [Getter/Setter],
      label: [Getter/Setter],
      description: [Getter/Setter],
      sortText: [Getter/Setter],
      canResolveChildren: [Getter/Setter],
      busy: [Getter/Setter],
      error: [Getter/Setter],
      tags: [Getter/Setter]
    }
  ],
  exclude: [],
  profile: b7 {
    controllerId: 'fsharp-test-controller',
    profileId: -1169225476,
    kind: 1,
    g: 'Run F# Tests',
    runHandler: [Function: runHandler],
    _tag: undefined,
    j: false
  },
  continuous: false,
  preserveFocus: true
}
[20:22:02 DEBUG] [TestExplorer] Test Filter Expression: (FullyQualifiedName~IcedTasks.AsyncEx.AsyncExBuilder.Return)|(FullyQualifiedName~IcedTasks.AsyncEx.AsyncExBuilder.Return)|(FullyQualifiedName~IcedTasks.AsyncEx.AsyncExBuilder.Return)
[20:22:05 DEBUG] [TestExplorer] Test run failed with exception Error: An item with the same key has already been added. Key: IcedTasks.AsyncEx.AsyncExBuilder.Return.Simple return
    at Dictionary__Add_5BDDA1 (c:\Users\jimmy\Repositories\public\TheAngryByrd\ionide-vscode-fsharp\release\webpack:\out\fable_modules\fable-library.4.2.2\MutableMap.js:273:1)
    at new Dictionary (c:\Users\jimmy\Repositories\public\TheAngryByrd\ionide-vscode-fsharp\release\webpack:\out\fable_modules\fable-library.4.2.2\MutableMap.js:19:1)
    at ArrayExt_venn (c:\Users\jimmy\Repositories\public\TheAngryByrd\ionide-vscode-fsharp\release\webpack:\out\Components\TestExplorer.js:43:37)
    at Interactions_mergeTestResultsToExplorer (c:\Users\jimmy\Repositories\public\TheAngryByrd\ionide-vscode-fsharp\release\webpack:\out\Components\TestExplorer.js:1430:1)
    at c:\Users\jimmy\Repositories\public\TheAngryByrd\ionide-vscode-fsharp\release\webpack:\out\Components\TestExplorer.js:1731:1
    at f2 (c:\Users\jimmy\Repositories\public\TheAngryByrd\ionide-vscode-fsharp\release\webpack:\out\fable_modules\fable-library.4.2.2\Util.js:572:1)
    at mergeResults (c:\Users\jimmy\Repositories\public\TheAngryByrd\ionide-vscode-fsharp\release\webpack:\out\Components\TestExplorer.js:1624:1)
    at c:\Users\jimmy\Repositories\public\TheAngryByrd\ionide-vscode-fsharp\release\webpack:\out\Components\TestExplorer.js:1700:1
```

TheAngryByrd commented:
Oh hell yes, being able to show multiple projects. ❤️ 

<img width="1317" height="257" alt="image" src="https://github.com/user-attachments/assets/dc4b38a1-11f0-472b-9879-71379781f4b4" />

farlee2121 commented:
> Thinking about that error. it's probably because I do have the same tests across different projects/tfms.

The regression testing repository I use has a project with multiple target frameworks, and it behaves as expected.

I did confirm that IcedTasks throws this error when I try to run all tests, but not when I try to run the projects one at a time. I'll need to dig deeper.

farlee2121 commented on src/Components/TestExplorer.fs:
```diff
@@ -79,6 +79,22 @@ module Dict =
+        option |> Option.iter f
+        option
+
+    let tryFallback f opt =
```
That's a handy core function to know about.

farlee2121 commented on src/Components/TestExplorer.fs:
```diff
@@ -220,6 +236,16 @@ type TestResultOutcome =
+    | Skipped
+
+module TestResultOutcome =
+    let ofOutcomeDto (outcomeDto: TestOutcomeDTO) =
```
I updated it to throw an error on an expected value.
My reasoning was that the only way we should have an unexpected error is if the language server changes.
In which case we'd want to know right away in development that we failed to sync the data contracts.

The downside, if a new test outcome value slips through it creates a more disruptive experience for effected users (though probably also a less confusing one than it mapping to another test outcome)

TheAngryByrd replied:
Since we control both sides of that DU I think it makes sense to fail early.

farlee2121 commented on src/Components/TestExplorer.fs:
```diff
@@ -1594,114 +2149,18 @@ module Interactions =
-
-                            warn
-                                $"No tests discovered for the following projects. Make sure your tests can be run with `dotnet test` \n {projectList}"
+                    do! discoverTests_WithLanguageServer testItemFactory rootTestCollection tryGetLocation
```
Sort of. I decided to have the full refresh still build the projects, because I'd be confused if I clicked refresh and my new tests didn't show up.

We could look at discovering tests without building when the project is initially loaded.

Eventually I'd like to move to tests being discovered whenever the build artifacts change.
That'd be more inline with the visual studio experience and narrows the gap in experience between projects that have code analysis-based update and those that don't.

farlee2121 replied:
I updated it to discover tests without building when first loading the workspace. 
It'll still build when the refresh button is clicked or when they run tests.

It feels much more responsive. The only potential loss is that the initial test list won't be up to date if the code has been changed since the last build. But the code will build and the explorer will update when tests are run.

To clarify, you want me to make an issue for watching build artifacts and responsively discovering tests?

farlee2121 commented:
@TheAngryByrd  I fixed the conflicting test name issue you encountered in IcedTasks and added a replication of the cause to my regression testing repo

TheAngryByrd commented:
Thanks for all the hard work on this!
